# Exploring Ebay Car Sales Data

In this project we will clean a dataset named [autos.csv](https://www.kaggle.com/orgesleka/used-cars-database/data) with 50,000 data points and then analyse the dataset.

In [90]:
import pandas as pd
import numpy as np
from stringcase import snakecase

In [91]:
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [92]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [93]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [94]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can observe that there are NULL values present in our dataset in some of the columns namely **vehicleType** , **gearbox** , **model** , **fuelType** and **notRepairedDamage**. Some of the columns have date and time but in string format.

In [95]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [96]:
col = autos.columns
col = col.str.replace("yearOfRegistration","registration_year")
col = col.str.replace("monthOfRegistration","registration_month")
col = col.str.replace("notRepairedDamage","unrepaired_damage")
col = col.str.replace("dateCreated","ad_created")
print(col)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [97]:
for name in col:
    col = col.str.replace(name,snakecase(name))
print(col)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [98]:
autos.columns = col

In [99]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above we renamed some of our columns and converted the names in camel case format to snake case format so that we dont face any issue during analysis.

In [100]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the above description we see that there are many columns that have only 2 unique values. There are 2 columns named **price** and **odometer** that have numeric data but is stored in object format.

In [101]:
autos["price"].describe()

count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object

In [102]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [103]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [104]:
autos["odometer"].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object

In [105]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [106]:
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos.rename({"odometer":"odometer_km"},axis = 1,inplace=True)

In [107]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0    manuell        158  andere       150000                   3       lpg   
1  automatik        286     7er       150000                   6    benzin   
2    manuell        102    golf        70000                   7    benzin   
3  automatik         71  fortwo        70000                   6    benzin   
4    manuell          0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [108]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000], dtype=int64)

In [109]:
autos["price"].unique()

array([ 5000.,  8500.,  8990., ...,   385., 22200., 16995.])

In [110]:
autos["price"].min()

0.0

In [111]:
autos["price"].max()

99999999.0

In [112]:
autos["odometer_km"].min()

5000

In [113]:
autos["odometer_km"].max()

150000

In [114]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [115]:
autos["price"].value_counts().sort_index().tail(25)

169999.0      1
175000.0      1
180000.0      1
190000.0      1
194000.0      1
197000.0      1
198000.0      1
220000.0      1
250000.0      1
259000.0      1
265000.0      1
295000.0      1
299000.0      1
345000.0      1
350000.0      1
999990.0      1
999999.0      2
1234566.0     1
1300000.0     1
3890000.0     1
10000000.0    1
11111111.0    2
12345678.0    3
27322222.0    1
99999999.0    1
Name: price, dtype: int64

In [116]:
autos["price"].value_counts().sort_index().head(25)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
40.0       6
45.0       4
47.0       1
49.0       4
50.0      49
Name: price, dtype: int64

So from above calculations, we can see that there are a lot of bad values present in price column of the dataset. We have 1421 data points with price of car as 0 dollars which is not possible. And also we see that there are few data points with random prices like 12345678 dollars or 99999999 dollars which is also not possible. So we need to remove quiet a good amount of data for a clean dataset. So lets set the outliner from 500 dollars  to 350000 dollars and keep data points whose price lie between this range only.

In [117]:
autos = autos[autos['price'].between(500.0,350000.0,inclusive = True)]

In [118]:
autos["price"].min()

500.0

In [119]:
autos["price"].max()

350000.0

In [120]:
autos.shape

(45097, 20)

So we can see that we removed almost 4900 rows from the dataset so that price remains in range 500 dollars to 350000 dollars.

In [121]:
#working with date data
autos["date_crawled"].describe()

count                   45097
unique                  43660
top       2016-03-09 11:54:38
freq                        3
Name: date_crawled, dtype: object

There are no missing values in date_crawled column

In [122]:
autos["ad_created"].describe()

count                   45097
unique                     76
top       2016-04-03 00:00:00
freq                     1761
Name: ad_created, dtype: object

There are no missing values in ad_created column

In [124]:
autos["last_seen"].describe()

count                   45097
unique                  35960
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

There are no missing values in last_seen column

In [134]:
autos["date_crawled"].str[:10].value_counts(normalize = True,dropna = False).sort_index(ascending = False)

2016-04-07    0.001353
2016-04-06    0.003171
2016-04-05    0.013172
2016-04-04    0.036610
2016-04-03    0.038827
2016-04-02    0.035767
2016-04-01    0.033905
2016-03-31    0.031665
2016-03-30    0.033328
2016-03-29    0.033262
2016-03-28    0.034836
2016-03-27    0.031177
2016-03-26    0.032641
2016-03-25    0.031089
2016-03-24    0.028982
2016-03-23    0.032397
2016-03-22    0.033018
2016-03-21    0.037741
2016-03-20    0.038073
2016-03-19    0.034747
2016-03-18    0.012883
2016-03-17    0.031155
2016-03-16    0.029359
2016-03-15    0.034016
2016-03-14    0.036300
2016-03-13    0.015522
2016-03-12    0.037320
2016-03-11    0.033018
2016-03-10    0.032707
2016-03-09    0.032907
2016-03-08    0.033173
2016-03-07    0.036189
2016-03-06    0.014125
2016-03-05    0.025567
Name: date_crawled, dtype: float64

We see that the website was crawled a lot of times between 25-03-2016 and 04-04-2016

In [136]:
autos["ad_created"].str[:10].value_counts(normalize = True,dropna = False).sort_index(ascending = False)

2016-04-07    0.001197
2016-04-06    0.003260
2016-04-05    0.011908
2016-04-04    0.036987
2016-04-03    0.039049
                ...   
2015-12-05    0.000022
2015-11-10    0.000022
2015-09-09    0.000022
2015-08-10    0.000022
2015-06-11    0.000022
Name: ad_created, Length: 76, dtype: float64

In [137]:
autos["last_seen"].str[:10].value_counts(normalize = True,dropna = False).sort_index(ascending = False)

2016-04-07    0.134155
2016-04-06    0.225314
2016-04-05    0.126616
2016-04-04    0.024303
2016-04-03    0.024946
2016-04-02    0.024880
2016-04-01    0.022862
2016-03-31    0.023438
2016-03-30    0.024148
2016-03-29    0.021354
2016-03-28    0.020534
2016-03-27    0.015456
2016-03-26    0.016476
2016-03-25    0.018582
2016-03-24    0.019536
2016-03-23    0.018405
2016-03-22    0.021243
2016-03-21    0.020667
2016-03-20    0.020423
2016-03-19    0.015411
2016-03-18    0.007406
2016-03-17    0.027674
2016-03-16    0.016165
2016-03-15    0.015677
2016-03-14    0.012285
2016-03-13    0.008870
2016-03-12    0.023904
2016-03-11    0.012041
2016-03-10    0.010289
2016-03-09    0.009468
2016-03-08    0.007007
2016-03-07    0.005211
2016-03-06    0.004169
2016-03-05    0.001087
Name: last_seen, dtype: float64

In [138]:
autos["registration_year"].describe()

count    45097.000000
mean      2005.064173
std         89.652017
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum registration year is 1000 and maximum is 9999 which is totally unrealistic. So we will remove all the unwanted data points with unrealistic dates. We will keep only data points lying between 1900 to 2016 and remove rest.

But before that lets first analyse the percentage of data we will be losing.

In [141]:
autos["registration_year"].between(1900,2016,inclusive = True).mean()

0.9606625717896977

So we see that we are not losing large amount of data. Only 4% of data are beyond that range. So we can easily remove them.

In [142]:
autos = autos[autos["registration_year"].between(1900,2016,inclusive = True)]

In [144]:
autos.shape

(43323, 20)

In [145]:
autos["registration_year"].describe()

count    43323.000000
mean      2003.227270
std          7.075229
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

So now we see that the minimum year is 1910 and maximum is 2016 which is quiet realistic. So now we can proceed with this data.

In [146]:
brands = autos["brand"].unique()

In [147]:
brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [149]:
autos["brand"].value_counts()

volkswagen        9180
bmw               4983
mercedes_benz     4430
opel              4347
audi              3921
ford              2823
renault           1902
peugeot           1272
fiat              1019
seat               775
skoda              743
smart              655
nissan             650
mazda              650
citroen            613
toyota             581
hyundai            440
sonstige_autos     420
mini               406
volvo              406
honda              338
mitsubishi         335
kia                318
alfa_romeo         287
porsche            278
chevrolet          261
suzuki             256
chrysler           156
dacia              123
jeep               106
land_rover          98
daihatsu            94
subaru              86
saab                72
jaguar              70
daewoo              58
rover               54
trabant             48
lancia              43
lada                26
Name: brand, dtype: int64

In [151]:
brand_percentage = autos["brand"].value_counts(normalize = True , ascending = False)

In [154]:
common_brand = brand_percentage[brand_percentage>0.05].index

In [155]:
common_brand

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford'], dtype='object')

In [156]:
price_aggregation = {}
for b in common_brand:
    rows = autos[autos["brand"]==b]
    mean = rows["price"].mean()
    price_aggregation[b]=mean
price_aggregation

{'volkswagen': 5783.622984749455,
 'bmw': 8582.261288380494,
 'mercedes_benz': 8766.902708803611,
 'opel': 3394.0395675178283,
 'audi': 9613.64779393012,
 'ford': 4291.666312433581}

From the above aggregation we can say that **Audi is the Costliest car** followed by BMW and Benz. Ford and Opel are less expensive whereas Volkawagen lies in between.

In [158]:
mean_mileage = {}
for b in common_brand:
    rows = autos[autos["brand"]==b]
    mean = rows["odometer_km"].mean()
    mean_mileage[b]=mean
mean_mileage

{'volkswagen': 128234.74945533769,
 'bmw': 132865.7435279952,
 'mercedes_benz': 131019.18735891648,
 'opel': 128012.42236024844,
 'audi': 128909.7169089518,
 'ford': 123494.50938717676}

In [160]:
mean_price = pd.Series(price_aggregation)

In [161]:
mileage = pd.Series(mean_mileage)

In [166]:
dataframe = pd.DataFrame(mean_price,columns = ["Mean Price"])

In [167]:
dataframe["Mileage"] = mileage

In [170]:
dataframe.sort_values("Mileage",ascending = False)

Mean Price        Mileage
bmw            8582.261288  132865.743528
mercedes_benz  8766.902709  131019.187359
audi           9613.647794  128909.716909
volkswagen     5783.622985  128234.749455
opel           3394.039568  128012.422360
ford           4291.666312  123494.509387

We see that Audi being the costliest car comes at 3rd rank in mileage whereas the car with highest mileage is BMW followed by Benz. Ford , Opel and Volkswagen have their mileages in ascending order respectively. 

# Conclusion:

We can conclude from the above analysis that expensive cars have comparatively high mileage compared to less expensive cars. Volkwagen being a car with moderate price i.e neither too expensive nor less expensive has its mileage as per its price. So price affects the mileage. Therefore if one is looking for cars with heavy mileage then he/she should go with Audi, BMW or Benz whereas people having low budget can go with Opel and Ford. And those who want affordable car with good mileage can go with **Volkswagen**.

## Thank You